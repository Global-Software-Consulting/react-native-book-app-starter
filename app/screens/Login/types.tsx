export interface ILoginData {
    email: string;
    password: string;
}
