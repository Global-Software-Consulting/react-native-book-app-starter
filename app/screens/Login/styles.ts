import AppStyles from 'config/styles';
import { StyleSheet } from 'react-native';
import { getPercentageWidth, getPercentageHeight } from 'utils/dimentionUtil';

export const useStyles = () => {
    const height = getPercentageHeight();
    const width = getPercentageWidth();
    const styles = StyleSheet.create({
        container: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
        },
        darkcontainer: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
        },
        guidingline: {
            fontSize: 15,
            fontWeight: '300',
            color: 'grey',
            alignSelf: 'center',
        },
        emailText: {
            width: width('80%'),
            height: height('6%'),
            padding: 5,
            alignSelf: 'center',
            borderWidth: 1,
            borderRadius: 20,
            marginTop: 10,
            backgroundColor: 'white',
        },

        passwordText: {},
        iconEye: {
            margin: 4,
        },
        loginText: {
            color: 'white',
        },
        activityIndicator: {
            margin: 5,
        },
        signupText: {
            color: '#db7093',
        },
        loginButton: {
            height: height('6 %'),
            width: width('60%'),
            marginTop: 5,
        },
        textMessage: {
            flexDirection: 'row',
            justifyContent: 'center',
            marginTop: 5,
        },
        touchableHighlight: {
            justifyContent: 'center',
            alignContent: 'center',
            flex: 1,
            alignItems: 'center',
        },
        forgetPasswordText: {
            alignSelf: 'flex-end',
            marginRight: 20,
            color: 'black',
            justifyContent: 'center',
        },
        loginView: {
            backgroundColor: '#491484',
            alignContent: 'center',
            alignSelf: 'center',
            justifyContent: 'space-between',
            flexDirection: 'row',
            marginTop: 15,
            height: height('5%'),
        },
        heading: {
            fontSize: 30,
            fontWeight: '500',
            alignSelf: 'center',
            justifyContent: 'center',
            marginTop: 20,
        },
        subcontainer: {
            // flex: 1,
            marginTop: '10%',
            width: '60%',
            height: '50%',
            justifyContent: 'center',
            alignSelf: 'center',
            alignItems: 'center',
            borderWidth: 0.25,
            borderColor: AppStyles.colors.COLOR_DARK_SEPERATOR,
        },
        login: {
            margin: 12,
        },

        forgot: {
            color: AppStyles.colors.COLOR_BLACK,
            marginTop: 12,
        },
        label: {
            fontSize: 12,
        },
        linearGradient: {
            height: height('25%'),
            alignItems: 'center',
            justifyContent: 'center',
            width: width('100%'),
        },
        welcomeText: { color: '#FAF9F6', fontSize: 30 },
        subHeading: {
            color: '#FAF9F6',
            fontSize: 15,
            maxWidth: width('85%'),
        },
        cardView: {
            backgroundColor: 'white',
            width: width('90%'),
            zIndex: 5,
            borderRadius: 20,
            marginTop: height('-2%'),
            alignSelf: 'center',
            justifyContent: 'center',
        },
        logo: { alignSelf: 'center', position: 'relative' },
        logInText: {
            fontSize: 30,
            fontWeight: '500',
            alignSelf: 'center',
            justifyContent: 'center',
            marginTop: 20,
            color: '#3A3B3C',
        },
        subLoginText: {
            fontSize: 15,
            fontWeight: '300',
            color: 'grey',
            alignSelf: 'center',
            marginBottom: height('3%'),
        },
        miniView: {
            fontSize: 15,
            fontWeight: '300',
            color: 'grey',
            alignSelf: 'center',
        },
        emailInput: {
            alignSelf: 'center',
            borderRadius: 20,
            margin: 5,
            backgroundColor: 'white',
            width: width('60%'),
        },
        passwordView: {
            flexDirection: 'row',
            borderColor: 'black',
            justifyContent: 'space-between',
            alignItems: 'center',
            width: width('62%'),
            height: 50,
            marginTop: 5,
            alignSelf: 'center',
        },
        passwordInput: {
            alignSelf: 'center',
            borderRadius: 20,
            margin: 5,
            backgroundColor: 'white',
            width: width('60%'),
        },
        icon: { position: 'absolute', marginLeft: '85%' },
        errorMessage: { alignSelf: 'center', color: 'red' },
        touchableOpacity: {
            justifyContent: 'center',
            alignContent: 'center',
            alignItems: 'center',
        },
        link: {
            alignSelf: 'flex-end',
            marginRight: 20,
            color: '#00416A',
            justifyContent: 'center',
        },
        submitView: {
            backgroundColor: '#00416A',
            alignContent: 'center',
            alignSelf: 'center',
            justifyContent: 'space-between',
            flexDirection: 'row',
            marginTop: 30,
            borderRadius: 20,
        },
        menu: {
            margin: 5,
            justifyContent: 'flex-end',
            alignItems: 'flex-end',
        },
        submit: {
            height: 40,
            width: width('60%'),
            marginTop: 5,
        },
        activity: {
            marginLeft: width('50%'),
            alignSelf: 'center',
            position: 'absolute',
        },
        basicText: {
            flexDirection: 'row',
            justifyContent: 'center',
            marginTop: 5,
        },
    });

    return styles;
};
