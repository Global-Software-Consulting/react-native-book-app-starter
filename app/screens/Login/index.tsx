//import { GoogleSignin } from '@react-native-google-signin/google-signin';
import { useNavigation } from '@react-navigation/core';
import { ReducerState } from 'models/reducers/index';
import React, { useState } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { ActivityIndicator, Image, Text, TextInput, TouchableOpacity, View } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import LinearGradient from 'react-native-linear-gradient';
import { Button } from 'react-native-paper';
import { Menu, MenuOption, MenuOptions, MenuTrigger } from 'react-native-popup-menu';
import Icon from 'react-native-vector-icons/FontAwesome';
import { useDispatch, useSelector } from 'react-redux';
import { UserData } from 'store/actions/loginActions';
import { getPercentageHeight, getPercentageWidth } from 'utils/dimentionUtil';
import images from './../../config/images';
import * as loginActions from './../../store/actions/loginActions';
import Toast from 'react-native-simple-toast';
import { useStyles } from './styles';
const Login: React.FC = () => {
    const dispatch = useDispatch();
    const height = getPercentageHeight();
    const width = getPercentageWidth();

    const navigation = useNavigation();
    const { t, i18n } = useTranslation();
    const styles = useStyles();
    const [secure, setSecure] = useState<boolean>(true);
    const [showActivityIndicator, setShowActivityIndicator] = useState<boolean>(false);
    const isLoading = useSelector((state: ReducerState) => state.loadingReducer.isLoading);
    //form data

    const {
        control,
        handleSubmit,
        formState: { errors },
    } = useForm();

    const validate = (text: string) => {
        let reg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w\w+)+$/;
        return reg.test(text);
    };

    const performLoginOperation = async (data: UserData) => {
        setShowActivityIndicator(isLoading);
        dispatch(loginActions.requestLogin(data));
    };

    const onSubmit = (data: UserData) => {
        if (validate(data.email)) {
            performLoginOperation(data);
        } else {
            Toast.show('Incorrect email format');
        }
    };

    // const window = Dimensions.get('window');
    // const screen = Dimensions.get('screen');
    // const [dimensions, setDimensions] = useState({ window, screen });

    // const loginWithGoogle = async () => {
    //     console.log('called');
    //     try {
    //         await GoogleSignin.hasPlayServices();
    //         const userInfo = await GoogleSignin.signIn();
    //         const x = JSON.stringify(userInfo);
    //         console.log(x);
    //         console.log('Google response', x);
    //     } catch (error) {
    //         if (error.code === statusCodes.SIGN_IN_CANCELLED) {
    //             // user cancelled the login flow
    //         } else if (error.code === statusCodes.IN_PROGRESS) {
    //             // operation (e.g. sign in) is in progress already
    //         } else if (error.code === statusCodes.PLAY_SERVICES_NOT_AVAILABLE) {
    //             // play services not available or outdated
    //         } else {
    //             // some other error happened
    //         }
    //     }
    // };
    function IconLanguage() {
        return (
            <View style={{ alignItems: 'center' }}>
                <Image source={images.app.language} />
                <Text>{i18n.language}</Text>
            </View>
        );
    }
    return (
        <KeyboardAwareScrollView style={{ flex: 1 }} showsVerticalScrollIndicator={false}>
            <LinearGradient
                colors={['#00416A', '#00416A', '#E4E5E6']}
                start={{ x: 0.0, y: 0.5 }}
                end={{ x: 0.1, y: 3.0 }}
                locations={[0, 0.5, 0.6]}
                style={styles.linearGradient}>
                <Text style={styles.welcomeText}>Welcome</Text>
                <Text style={styles.subHeading}>
                    Book app aims to provide variety with a quick sharing of resources among friends
                    and family.
                </Text>
            </LinearGradient>
            <View style={styles.cardView}>
                <Image source={images.app.logo} style={styles.logo} />

                <Text style={styles.logInText}>{t('Log In Now')}</Text>

                <Text style={styles.subLoginText}>
                    {t('Please login to continue using our app')}
                </Text>
                <View style={styles.miniView}>
                    <Controller
                        control={control}
                        rules={{
                            required: true,
                        }}
                        render={({ field: { onChange, value } }) => (
                            <TextInput
                                placeholder={t('Enter your email address')}
                                autoCapitalize="none"
                                textContentType="emailAddress"
                                keyboardType="email-address"
                                autoCorrect={false}
                                style={styles.emailInput}
                                value={value}
                                onChangeText={(text) => onChange(text)}
                            />
                        )}
                        name="email"
                        defaultValue=""
                    />
                    {errors.email && (
                        <Text style={{ alignSelf: 'center', color: 'red' }}>
                            {t('Email is required')}
                        </Text>
                    )}

                    <View style={styles.passwordView}>
                        <Controller
                            control={control}
                            rules={{
                                required: true,
                            }}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    placeholder={t('Enter your password')}
                                    autoCapitalize="none"
                                    secureTextEntry={secure}
                                    autoCorrect={false}
                                    style={styles.passwordInput}
                                    value={value}
                                    onChangeText={(text) => onChange(text)}
                                />
                            )}
                            name="password"
                            defaultValue=""
                        />
                        <Icon
                            name={secure ? 'eye-slash' : 'eye'}
                            size={22}
                            style={styles.icon}
                            onPress={() => setSecure(!secure)}
                        />
                    </View>
                    {errors.password && (
                        <Text style={styles.errorMessage}>{t('Password is required')}</Text>
                    )}
                </View>

                <TouchableOpacity
                    style={styles.touchableOpacity}
                    onPress={() => navigation.navigate('ForgotPassword' as never)}>
                    <Text
                        style={styles.link}
                        onPress={() => navigation.navigate('ForgotPassword' as never)}>
                        {t('Forgot Password')}
                    </Text>
                </TouchableOpacity>

                <View style={styles.submitView}>
                    <Button
                        onPress={handleSubmit(onSubmit)}
                        disabled={showActivityIndicator}
                        style={styles.submit}>
                        <Text style={{ color: 'white' }}>{t('Log in')}</Text>
                    </Button>
                    {isLoading && <ActivityIndicator color="white" style={styles.activity} />}
                </View>

                <View style={styles.basicText}>
                    <Text>{t('Do not have an account?')} </Text>
                    <Text
                        style={{ color: '#00416A' }}
                        onPress={() => navigation.navigate('Signup' as never)}>
                        {t('Sign up')}
                    </Text>
                </View>

                <View style={styles.menu}>
                    <Menu>
                        <MenuTrigger>
                            <IconLanguage />
                        </MenuTrigger>
                        <MenuOptions>
                            <MenuOption
                                onSelect={() => {
                                    i18n.changeLanguage('en');
                                }}
                                text="English"
                            />
                            <MenuOption
                                onSelect={() => {
                                    i18n.changeLanguage('es');
                                }}
                                text="Spanish"
                            />
                            <MenuOption
                                onSelect={() => {
                                    i18n.changeLanguage('de');
                                }}
                                text="Dutch"
                            />
                        </MenuOptions>
                    </Menu>
                </View>
            </View>
        </KeyboardAwareScrollView>
    );
};

export default React.memo(Login);
