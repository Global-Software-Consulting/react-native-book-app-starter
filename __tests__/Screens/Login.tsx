import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { useDispatch } from 'react-redux'
import { act } from 'react-test-renderer'
import Login from './../../app/screens/Login/index'
jest.useFakeTimers()

jest.mock('react-redux', () => {
    return {
        __esModule: true,
        A: true,
        useDispatch: jest.fn(),
        useSelector: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('@react-navigation/core', () => {
    return {
        __esModule: true,
        A: true,
        useNavigation: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('./../../app/config/images', () => {
    return {
        __esModule: true,
        A: true,
        images: {
            app: {
                logo: 'abc',
            },
        },
        default: 'mockedDefaultExport',
    }
})
jest.mock('react-native-keyboard-aware-scroll-view', () => {
    return {
        __esModule: true,
        A: true,
        KeyboardAwareScrollView: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('react-native-linear-gradient', () => {
    return {
        __esModule: true,
        A: true,
        LinearGradient: jest.fn(),
        default: 'mockedDefaultExport',
    }
})

describe('Component testing', () => {
    test('Snapshot', async () => {
        const tree = render(<Login />)
        expect(tree).toMatchSnapshot()
    })
    test('submit button with wrong email and password ', async () => {
        const dispatch = jest.fn()
        useDispatch.mockReturnValue(jest.fn())
        const { getByTestId } = render(<Login />)
        fireEvent.changeText(getByTestId('email'), 'dsa')
        fireEvent.changeText(getByTestId('password'), 'fdsa')

        fireEvent.press(getByTestId('submit'))
    })

    test('submit button with email and password ', async () => {
        const dispatch = jest.fn()
        useDispatch.mockReturnValue(jest.fn())
        const { getByTestId } = render(<Login />)
        fireEvent.changeText(getByTestId('email'), 'dsa@gmail.com')
        fireEvent.changeText(getByTestId('password'), 'fdsa')

        fireEvent.press(getByTestId('submit'))
    })
    test('submit button ', async () => {
        const { getByTestId } = render(<Login />)
        act(() => {
            fireEvent.press(getByTestId('submit'))
        })
    })

    test('icon press event ', async () => {
        const { getByTestId } = render(<Login />)
        act(() => {
            fireEvent.press(getByTestId('iconEye'))
        })
    })
})
