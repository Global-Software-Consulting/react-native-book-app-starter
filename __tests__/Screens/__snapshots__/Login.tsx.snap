// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component testing Snapshot 1`] = `
Array [
  <mockedDefaultExport
    colors={
      Array [
        "#00416A",
        "#00416A",
        "#E4E5E6",
      ]
    }
    end={
      Object {
        "x": 0.1,
        "y": 3,
      }
    }
    locations={
      Array [
        0,
        0.5,
        0.6,
      ]
    }
    start={
      Object {
        "x": 0,
        "y": 0.5,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "height": 333.5,
        "justifyContent": "center",
        "width": 750,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FAF9F6",
          "fontSize": 30,
        }
      }
    >
      Welcome
    </Text>
    <Text
      style={
        Object {
          "color": "#FAF9F6",
          "fontSize": 15,
          "maxWidth": 637.5,
        }
      }
    >
      Book app aims to provide variety with a quick sharing of resources among friends and family.
    </Text>
  </mockedDefaultExport>,
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "justifyContent": "center",
        "marginTop": -26.5,
        "width": 675,
        "zIndex": 5,
      }
    }
  >
    <Image
      style={
        Object {
          "alignSelf": "center",
          "position": "relative",
        }
      }
    />
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#3A3B3C",
          "fontSize": 30,
          "fontWeight": "500",
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    />
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "grey",
          "fontSize": 15,
          "fontWeight": "300",
          "marginBottom": 40,
        }
      }
    />
    <View
      style={
        Object {
          "alignSelf": "center",
          "color": "grey",
          "fontSize": 15,
          "fontWeight": "300",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        keyboardType="email-address"
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 20,
            "margin": 5,
            "width": 450,
          }
        }
        testID="email"
        textContentType="emailAddress"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "space-between",
            "marginTop": 5,
            "width": 465,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          onChangeText={[Function]}
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "margin": 5,
              "width": 450,
            }
          }
          testID="password"
          underlineColorAndroid="transparent"
          value=""
        />
        <mockedDefaultExport
          name="eye-slash"
          onPress={[Function]}
          size={22}
          style={
            Object {
              "marginLeft": "85%",
              "position": "absolute",
            }
          }
          testID="iconEye"
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID="forgotPassword"
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#00416A",
            "justifyContent": "center",
            "marginRight": 20,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignContent": "center",
          "alignSelf": "center",
          "backgroundColor": "#00416A",
          "borderRadius": 20,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 30,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 0,
            "height": 40,
            "marginTop": 5,
            "minWidth": 64,
            "width": 450,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
          testID="submit"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#6200ee",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                  }
                }
              />
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 5,
        }
      }
    >
      <Text>
         
      </Text>
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#00416A",
          }
        }
        testID="signup"
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "justifyContent": "flex-end",
          "margin": 5,
        }
      }
    >
      <Menu>
        <MenuTrigger>
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Image />
            <Text />
          </View>
        </MenuTrigger>
        <MenuOptions>
          <MenuOption
            onSelect={[Function]}
            text="English"
          />
          <MenuOption
            onSelect={[Function]}
            text="Spanish"
          />
          <MenuOption
            onSelect={[Function]}
            text="Dutch"
          />
        </MenuOptions>
      </Menu>
    </View>
  </View>,
]
`;
